### PR TITLE
Fix LQP outputs

### PR DESCRIPTION
**Pull request by mrks**
With both visit_lqp_upwards and the new ColumnPruningRule, we depend on the output links of AbstractLQPNodes to be correct. We now make sure that the optimizer has exclusive ownership so that no remnants of the old plan remain.

This also adds a verify_lqp method to the optimizer (which we should have had a long time ago). While it is not perfect, it should help us catch issues as soon as a rule breaks something, not just once a following rule or the LQPTranslator stumbles upon the issue.

TODO
- [x] Check impact on CI runtime
- [x] Check impact on benchmark results
- [x] Tests for Optimizer::verify_lqp

### CODE REVIEW

File: src/lib/logical_query_plan/lqp_utils.cpp [hunk 1]
```diff
@@ -296,4 +296,31 @@ std::vector<std::shared_ptr<AbstractLQPNode>> lqp_find_subplan_roots(const std::
   return root_nodes;
 }
 
+void collect_subquery_expressions_by_lqp(SubqueryExpressionsByLQP& subquery_expressions_by_lqp,
```
Comment (mrks) [inline src/lib/logical_query_plan/lqp_utils.cpp]: Moved here 1:1 from optimizer.cpp

File: src/lib/optimizer/optimizer.cpp [hunk 2]
```diff
@@ -148,6 +110,16 @@ std::shared_ptr<AbstractLQPNode> Optimizer::optimize(const std::shared_ptr<Abstr
...
   for (const auto& rule : _rules) {
     _apply_rule(*rule, root_node);
+
+    // The SubqueryToJoinRule was written before we required all outputs of a node to be part of the same plan
+    // (see validate_lqp). Other than that, the rule works correctly and refactoring it only for this additional
+    // requirements is not worth the effort. Thus, we manually remove all outputs that become unused due to the
+    // rewrite. We trigger this here so that it is only called on the root node, not on any subquery LQPs.
+    if (const auto subquery_to_join_rule = dynamic_cast<SubqueryToJoinRule*>(&*rule)) {
+      subquery_to_join_rule->remove_unused_outputs(root_node);
+    }
```
Comment (mrks) [inline src/lib/optimizer/optimizer.cpp]: Really unhappy about this part

**Comment by mrzzzrm** [comment]
The LQP (both in the input setters and the node destructor) unlinks itself, iirc? Under which circumstances does this not work?

**Comment by mrks** [comment]
Oh, good thing that you asked. I though I remembered that we decided against LQP nodes automatically untying themselves after destruction. That, however, does happen.

While nodes from within the LQP get correctly destructed, the caller might hold the original root node, which might not even be part of the LQP anymore.

Fixing that is MUCH cleaner than what I had before.

File: src/lib/logical_query_plan/lqp_column_reference.cpp [hunk 3]
```diff
@@ -26,10 +26,10 @@ std::ostream& operator<<(std::ostream& os, const LQPColumnReference& column_refe
   const auto original_node = column_reference.original_node();
   Assert(original_node, "OriginalNode has expired");
 
-  const auto stored_table_node = std::static_pointer_cast<const StoredTableNode>(column_reference.original_node());
+  const auto stored_table_node = std::dynamic_pointer_cast<const StoredTableNode>(column_reference.original_node());
```
Comment (Bensk1) [inline src/lib/logical_query_plan/lqp_column_reference.cpp]: Why don't we need to check the result of the cast for a `nulltpr`?

File: src/lib/optimizer/optimizer.cpp [hunk 4]
```diff
@@ -141,13 +141,23 @@ void Optimizer::add_rule(std::unique_ptr<AbstractRule> rule) {
   _rules.emplace_back(std::move(rule));
 }
 
-std::shared_ptr<AbstractLQPNode> Optimizer::optimize(const std::shared_ptr<AbstractLQPNode>& input) const {
+std::shared_ptr<AbstractLQPNode> Optimizer::optimize(std::shared_ptr<AbstractLQPNode> input) const {
+  // We cannot allow multiple owners of the LQP as one owner could decide to optimize the plan and others might hold a
+  // pointer to a node that is not even part of the plan anymore after optimization. Thus, callers of this method need
+  // to relinquish their ownership (i.e., move their shared_ptr into the method) and take ownership of the resulting
+  // optimized plan.
+  Assert(input.use_count() == 1, "Optimizer should have exclusive ownership of plan");
```
Comment (Bensk1) [inline src/lib/optimizer/optimizer.cpp]: Hm, are there any other ways than checking the use_count of the shared pointer to achieve this? This does not seem very clean.

File: src/lib/optimizer/optimizer.cpp [hunk 5]
```diff
@@ -157,6 +167,100 @@ std::shared_ptr<AbstractLQPNode> Optimizer::optimize(const std::shared_ptr<Abstr
...
+  // (1) Make sure that all outputs found anywhere in the entire plan are also part of the plan
+  // (2) Make sure each node has the number of inputs expected for that node type
+  // (3) Make sure that for all LQPColumnExpressions, the original_node is part of the LQP
+  for (const auto& lqp : lqps) {
+    visit_lqp(lqp, [&all_nodes](const auto& node) {
+      // Check that all outputs are part of the LQP
+      const auto outputs = node->outputs();
+      for (const auto& output : outputs) {
+        // LQP nodes should only be part of a single LQP and all of their outputs should be part of the same LQP. If you
+        // run into this assertion as part of a new test, there is a good chance that you are using part of the LQP
```
Comment (Bensk1) [inline src/lib/optimizer/optimizer.cpp]: partS or a part?

File: src/lib/optimizer/optimizer.cpp [hunk 6]
```diff
@@ -157,6 +167,100 @@ std::shared_ptr<AbstractLQPNode> Optimizer::optimize(const std::shared_ptr<Abstr
...
+        // outside of the LQP that is currently optimized. For example, if you create a bunch of LQP nodes in the test's
+        // SetUp method but these are not used in the current LQP, this can cause this assertion to fail. This is only
+        // enforced when rules are executed through the optimizer, not tests call the rule directly.
```
Comment (Bensk1) [inline src/lib/optimizer/optimizer.cpp]: not when tests

File: src/lib/sql/sql_pipeline.cpp [hunk 7]
```diff
@@ -145,16 +145,19 @@ const std::vector<std::shared_ptr<AbstractLQPNode>>& SQLPipeline::get_optimized_
          "One or more SQL statement is dependent on the execution of a previous one. "
          "Cannot translate all statements without executing, i.e. calling get_result_table()");
 
+  // The optimizer works on the original unoptimized LQP nodes. After optimizing, the unoptimized version is also
```
Comment (Bensk1) [inline src/lib/sql/sql_pipeline.cpp]: I think I understand what you want to say but I find it a bit confusing. Maybe it gets clearer if we just say that the unoptimized version is altered during the optimization process. 
A different question, why do we have to modify the unoptimized version at all, can't we just create a copy if we have `_optimized_logical_plans` anyways?

File: src/test/optimizer/optimizer_test.cpp [hunk 8]
```diff
@@ -43,6 +43,90 @@ class OptimizerTest : public ::testing::Test {
...
+      if (!projection_node) return;
+      projection_node->node_expressions[0] = out_of_plan_expression;
+    }
+
+    std::shared_ptr<AbstractExpression> out_of_plan_expression;
+  };
+
+  optimizer.add_rule(std::make_unique<LQPBreakingRule>(lqp_column_(x)));
+
+  // Does not move the LQP into the optimizer
```
Comment (Bensk1) [inline src/test/optimizer/optimizer_test.cpp]: Comment copy pasta?

File: src/test/optimizer/optimizer_test.cpp [hunk 9]
```diff
@@ -142,7 +228,8 @@ TEST_F(OptimizerTest, OptimizesSubqueriesExactlyOnce) {
   Optimizer optimizer{};
   optimizer.add_rule(std::move(rule));
 
-  optimizer.optimize(lqp);
+  const auto optimized_lqp = optimizer.optimize(std::move(lqp));
+  lqp = nullptr;
```
Comment (Bensk1) [inline src/test/optimizer/optimizer_test.cpp]: Why do we specifically set `lqp = nullptr` here and don't do that above?

File: src/lib/optimizer/optimizer.cpp (hunk 4 above)
Comment (mrks) [reply src/lib/optimizer/optimizer.cpp]: Could you try to explain your concerns?

File: src/lib/sql/sql_pipeline.cpp (hunk 7 above)
Comment (mrks) [reply src/lib/sql/sql_pipeline.cpp]:
```
// The optimizer modifies the input LQP and requires exclusive ownership of that LQP. This means that we need to clear
// _unoptimized_logical_plans. This is not an issue as the unoptimized plans will no longer be needed. Note that this
// does not clear the unoptimized LQPs stored in the SQLPipelineStatement - those are cleared as part of
// SQLPipelineStatement::get_optimized_logical_plan. Calls to get_unoptimized_logical_plans are still allowed (e.g.,
// for visualization), in which case the unoptimized plan will be recreated.
```

better?

Comment (mrks) [reply src/lib/sql/sql_pipeline.cpp]: If yes, I'll adapt the version in sql_pipeline_statement.cpp accordingly

File: src/test/optimizer/optimizer_test.cpp (hunk 9 above)
Comment (mrks) [reply src/test/optimizer/optimizer_test.cpp]: Mmh, no real reason other than that there were more than a couple of lines below. In the end, both should work, so I would just leave both versions.

File: src/lib/logical_query_plan/lqp_column_reference.cpp (hunk 3 above)
Comment (mrks) [reply src/lib/logical_query_plan/lqp_column_reference.cpp]: I was surprised to see only StoredTableNode here. For testing, I changed it to check for segfaults. In regular execution, we do not print MockNode expressions, but I fixed it anyway.

File: src/lib/optimizer/optimizer.cpp (hunk 4 above)
Comment (Bensk1) [reply src/lib/optimizer/optimizer.cpp]: As discussed offline

Comment (mrks) [reply src/lib/optimizer/optimizer.cpp]: Did we reach a conclusion?

File: src/lib/sql/sql_pipeline.cpp (hunk 7 above)
Comment (Bensk1) [reply src/lib/sql/sql_pipeline.cpp]: Yes

File: src/lib/optimizer/optimizer.cpp (hunk 4 above)
Comment (Bensk1) [reply src/lib/optimizer/optimizer.cpp]: It seems to be fine.

**Pull request merged**